File: Data Structure/Linked List/Hard/LET_no460_LFU Cache.py
class Node:
    def __init__(self, key, val):
        self.key, self.val = key, val
        self.prev = self.next = None
        self.freq = 1

class DList:
    def __init__(self):
        self.dummy = Node(-1, -1)
        self.dummy.prev = self.dummy.next = self.dummy
        self.size = 0
    
    def __len__(self):
        return self.size
    
    def append(self, node):
        tail = self.dummy.prev
        tail.next, node.prev = node, tail
        node.next, self.dummy.prev = self.dummy, node
        self.size += 1
    
    def pop(self, node=None):
        if len(self) == 0:
            return
        if not node:
            node = self.dummy.next
            
        prev_node, next_node = node.prev, node.next
        prev_node.next, next_node.prev = next_node, prev_node
        self.size -= 1
        
        return node

class LFUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.key_to_node = dict()
        self.freq_to_list = collections.defaultdict(DList)
        self.min_freq = 0

    def get(self, key: int) -> int:
        if key not in self.key_to_node:
            return -1
        
        node = self.key_to_node[key]
        self.updateFreq(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if self.capacity == 0:
            return
        if key in self.key_to_node:
            node = self.key_to_node[key]
            self.updateFreq(node)
            node.val = value
            return
        
        if len(self.key_to_node) == self.capacity:
            node = self.freq_to_list[self.min_freq].pop()
            del self.key_to_node[node.key]
            
        node = Node(key, value)
        self.key_to_node[key] = node
        self.freq_to_list[1].append(node)
        self.min_freq = 1
    
    def updateFreq(self, node):
        freq = node.freq
        self.freq_to_list[freq].pop(node)
        if freq == self.min_freq and len(self.freq_to_list[freq]) == 0:
            self.min_freq += 1
        
        freq += 1
        node.freq = freq
        self.freq_to_list[freq].append(node)


# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: Data Structure/Linked List/Hard/LET_no460_LFU Cache.cpp
class LFUCache {
private:
    int capacity, size, minFreq;
    unordered_map<int, list<int>> freq2keys;
    unordered_map<int, pair<int, int>> key2value;  // (value, freq)
    unordered_map<int, list<int>::iterator> key2pos;

public:
    LFUCache(int capacity) {
        this->capacity = capacity;
        size = minFreq = 0;
    }
    
    int get(int key) {
        if (!key2value.count(key))
            return -1;
        
        updateFreq(key);
        return key2value[key].first;
    }
    
    void put(int key, int value) {
        if (!capacity)
            return;
        
        if (key2value.count(key)) {
            key2value[key].first = value;
            updateFreq(key);
        }
        else {
            if (size == capacity) {
                int victimKey = freq2keys[minFreq].front();
                freq2keys[minFreq].pop_front();
                key2value.erase(victimKey);
                key2pos.erase(victimKey);
            }
            else
                size++;
            
            key2value[key] = {value, 1};
            freq2keys[1].push_back(key);
            key2pos[key] = --freq2keys[1].end();
            minFreq = 1;
        }
    }

    void updateFreq(int key) {
        int freq = key2value[key].second;
        auto iter = key2pos[key];
        key2value[key].second++;
        freq2keys[freq].erase(iter);
        freq2keys[freq + 1].push_back(key);
        key2pos[key] = --freq2keys[freq + 1].end();
        
        if (freq2keys[minFreq].empty())
            minFreq++;
    }
};

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache* obj = new LFUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */